[SR] Allow custom index pattern to be used instead of selectable list when choosing indices to restore

jen-huang opened the pull request:
## Summary
This PR is an enhancement to #39193. The original work overlooked that fact that when restoring a snapshot, an index pattern can be used to choose which indices to restore instead of index names. This is useful for users with with a large amount of indices and/or time series indices.

With this PR, there is a now a toggle link when the user turns off All Indices. By default, if the snapshot to restore has <= 100 indices, the selectable list will display. If the snapshot to restore has > 100 indices, the custom pattern input will display. **The custom pattern input does not do validation against the available indices**, so ES error will surface on Review page if it is invalid.

## Screenshots

**Selectable list input with toggle link to custom pattern input**

![image](https://user-images.githubusercontent.com/1965714/61497350-fb106680-a973-11e9-876a-833f43de10b0.png)

**Custom pattern input with toggle link to selectable list input**

![image](https://user-images.githubusercontent.com/1965714/61497409-22673380-a974-11e9-911b-3f5b0c5a5f9e.png)

**Review page using custom pattern input value `rollup*,kibana*,test_index`**

![image](https://user-images.githubusercontent.com/1965714/61497541-a9b4a700-a974-11e9-834f-542b68f416fd.png)
![image](https://user-images.githubusercontent.com/1965714/61497608-e7b1cb00-a974-11e9-909b-e9f505ab20c6.png)

**Review page surfacing an error with invalid index name in custom pattern input**

![image](https://user-images.githubusercontent.com/1965714/61497515-85f16100-a974-11e9-9023-f453d76c366b.png)

elasticmachine commented:
Pinging @elastic/es-ui

jen-huang commented on x-pack/legacy/plugins/snapshot_restore/public/app/sections/home/repository_list/repository_details/repository_details.tsx:
```diff
@@ -294,8 +294,8 @@ export const RepositoryDetails: React.FunctionComponent<Props> = ({
-              id="xpack.snapshotRestore.repositoryDetails.reverifyButtonLabel"
-              defaultMessage="Re-verify repository"
+              id="xpack.snapshotRestore.repositoryDetails.verifyButtonLabel"
+              defaultMessage="Verify repository"
```
this is a small copy change @gchaps requested

jen-huang commented:
@gchaps Would you please review the copy for the toggle links and the validation error message? Thanks!

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8339/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/8355/)

gchaps commented:
We typically use this format for validation on fields:

An index is required.

Make sure this format is consistent with the rest of the UI (i.e. that the validation on the Name field is "Name is required."_

Also, for the message on the review page, is it possible to change the heading to "Unable to restore snapshot."? I'd prefer not to use the word "execute".

Do we need to update the [S & R documentation](https://www.elastic.co/guide/en/kibana/7.3/snapshot-repositories.html)?

jen-huang commented:
@gchaps Thanks for the feedback!

The SR doc link doesn't seem to include detailed information about the Restore wizard, so I don't think that needs to be updated. I do remember we worked on a detailed walkthrough for the wizard that includes screenshots of all steps - will that be published? If so, there is probably 1 screenshot that should be updated.

alisonelizabeth commented on x-pack/legacy/plugins/snapshot_restore/public/app/components/restore_snapshot_form/steps/step_logistics.tsx:
```diff
@@ -156,89 +171,146 @@ export const RestoreSnapshotStepLogistics: React.FunctionComponent<StepProps> =
+                      )}
+                    </EuiSelectable>
+                  ) : (
+                    <EuiFieldText
```
What do you think about using the `EuiComboBox` here instead?

jen-huang replied:
I think this is a good idea, I'll play around with the placeholder text to make sure the user understands the kind of patterns they can enter

alisonelizabeth commented on x-pack/legacy/plugins/snapshot_restore/public/app/components/restore_snapshot_form/steps/step_logistics.tsx:
```diff
@@ -156,89 +171,146 @@ export const RestoreSnapshotStepLogistics: React.FunctionComponent<StepProps> =
-                          ),
-                      }}
-                    />
+                    selectIndicesMode === 'list' ? (
```
I don't have a good alternative atm, but i'm not sure if it would have been immediately clear to me that the `Custom pattern`/`Show indices` link is used to toggle views if i wasn't familiar with the PR changes.

jen-huang replied:
@gchaps Could you help me with the wording for the toggle links?

Currently they are: `Custom pattern` and `Select indices`. You can see them in the screenshots in PR description. I agree with Alison that the current wording doesn't correlate very strongly to the fact that the links themselves toggle between a text input view and a selectable list view.

Maybe they could be: `Enter custom pattern` and `Select from list` instead? Help!! 🙂

cjcenizal replied:
I had the same thought when I looked at the copy. How about "Use index pattern" and "Choose individual indices"?

In terms of prior art, in the Rollup Job Wizard we use "Create cron expression" and "Create basic interval" to switch between two UIs for creating what eventually becomes a cron expression, though I'm not convinced "Create" is the right verb to use here. I think I would prefer "Use".

![image](https://user-images.githubusercontent.com/1238659/61826489-59779200-ae17-11e9-826e-1f782e7652f3.png)

![image](https://user-images.githubusercontent.com/1238659/61826493-5aa8bf00-ae17-11e9-98e9-6c8298ab4f66.png)

gchaps replied:
@jen-huang Based on the suggestion from @cjcenizal , what about:

Use index pattern
Select index

cjcenizal replied:
Thanks @gchaps! Is there a reason to use "Select index" instead of "Select indices"? Is singular better than plural in these cases?

gchaps replied:
@cjcenizal If the user can select more than one index, then "Select indices" is better.

jen-huang commented:
Updated toggle link and validation copies, and changed index pattern input to combo box:

![image](https://user-images.githubusercontent.com/1965714/61765692-fcd29380-ad92-11e9-9063-bd4ad9f5e2fe.png)

![image](https://user-images.githubusercontent.com/1965714/61765674-e298b580-ad92-11e9-8682-6e7b8d9c8dc4.png)

![image](https://user-images.githubusercontent.com/1965714/61765680-f0e6d180-ad92-11e9-81f0-a4a46fbbd28a.png)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/9047/)

cjcenizal commented on x-pack/legacy/plugins/snapshot_restore/public/app/components/restore_snapshot_form/steps/step_logistics.tsx:
```diff
@@ -55,11 +56,25 @@ export const RestoreSnapshotStepLogistics: React.FunctionComponent<StepProps> =
 
+  // State for using selectable indices list or custom patterns
+  const [selectIndicesMode, setSelectIndicesMode] = useState<'list' | 'custom'>(
+    typeof restoreIndices === 'string' || snapshotIndices.length > 100 ? 'custom' : 'list'
```
I love the UX consideration we're making here. I think it warrants a comment, e.g.

```js
// Users with more than 100 indices will probably want to use an index pattern to select
// them instead, so we'll default to showing them the index pattern input.
```

Also, the Index Pattern Wizard shows you a preview of matching indices to your index pattern. Do you think users would benefit from something similar here? Even if so, I would not consider it a blocker for this PR.

cjcenizal replied:
BTW if we do decide to fetch matching indices at some point, https://github.com/elastic/kibana/issues/42045 contains a suggestion on how to make that request more efficient.

cjcenizal commented on x-pack/legacy/plugins/snapshot_restore/public/app/components/restore_snapshot_form/steps/step_logistics.tsx:
```diff
@@ -156,89 +174,158 @@ export const RestoreSnapshotStepLogistics: React.FunctionComponent<StepProps> =
-                      options.forEach(({ label, checked }) => {
-                        if (checked === 'on') {
-                          newSelectedIndices.push(label);
+                  {selectIndicesMode === 'list' ? (
```
Minor nit: we have three parallel branches that depend upon `selectIndicesMode`, do you think it would make sense to reduce this to a single branch that switches between an `IndicesSelect` component and an `IndexPatternInput` component? That would also reduce the size of this file, which is getting lengthy.

cjcenizal commented on x-pack/legacy/plugins/snapshot_restore/public/app/components/restore_snapshot_form/steps/step_logistics.tsx:
```diff
@@ -156,89 +174,158 @@ export const RestoreSnapshotStepLogistics: React.FunctionComponent<StepProps> =
+                  ) : (
+                    <EuiComboBox
+                      options={snapshotIndices.map(index => ({ label: index }))}
+                      placeholder="Enter index patterns or indices, i.e. logstash-*"
```
I think this needs to be internationalized.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10121/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/10129/)
